### PR TITLE
src: Throw proper expections instead of just a string

### PR DESCRIPTION
**Pull request by xdustinface**
~~Based on #4~~

### CODE REVIEW

**Comment by PastaPastaPasta** [comment]
Should this be done upstream?

**Comment by xdustinface** [comment]
Does upstream still maintain V1?

**Comment by PastaPastaPasta** [comment]
Did this change in v2?

**Comment by PastaPastaPasta** [comment]
see https://github.com/Chia-Network/bls-signatures/blob/master/src/bls.cpp

These changes can be applied there

**Comment by xdustinface** [comment]
Ok yeah partially it could be applied it seems but not for the dash only stuff. And for where it can be applied it doesn't matter that much because its only init means only called once on startup.

File: src/bls.cpp [hunk 1]
```diff
@@ -31,28 +31,22 @@ Util::SecureFreeCallback Util::secureFreeCallback;
 static void relic_core_initializer(void* ptr) {
     core_init();
     if (err_get_code() != STS_OK) {
-        std::cout << "core_init() failed";
-        // this will most likely crash the application...but there isn't much we can do
-        throw std::string("core_init() failed");
```
Comment (PastaPastaPasta) [inline src/bls.cpp:36]: This is the same upstream

File: src/bls.cpp [hunk 2]
```diff
@@ -31,28 +31,22 @@ Util::SecureFreeCallback Util::secureFreeCallback;
...
+        throw std::runtime_error("core_init() failed");
     }
 
     const int r = ep_param_set_any_pairf();
     if (r != STS_OK) {
-        std::cout << "ep_param_set_any_pairf() failed";
-        // this will most likely crash the application...but there isn't much we can do
-        throw std::string("ep_param_set_any_pairf() failed");
```
Comment (PastaPastaPasta) [inline src/bls.cpp:43]: same as upstream

File: src/bls.cpp [hunk 3]
```diff
@@ -31,28 +31,22 @@ Util::SecureFreeCallback Util::secureFreeCallback;
...
+        throw std::runtime_error("ep_param_set_any_pairf() failed");
     }
 }
 
 bool BLS::Init() {
     if (ALLOC != AUTO) {
-        std::cout << "Must have ALLOC == AUTO";
-        throw std::string("Must have ALLOC == AUTO");
```
Comment (PastaPastaPasta) [inline src/bls.cpp:50]: Same upstream

File: src/bls.cpp [hunk 4]
```diff
@@ -31,28 +31,22 @@ Util::SecureFreeCallback Util::secureFreeCallback;
...
+        throw std::runtime_error("Must have ALLOC == AUTO");
     }
 #if BLSALLOC_SODIUM
     if (sodium_init() < 0) {
-        std::cout << "libsodium init failed";
-        throw std::string("libsodium init failed");
```
Comment (PastaPastaPasta) [inline src/bls.cpp:55]: same upstream

File: src/bls.cpp [hunk 5]
```diff
@@ -314,7 +308,7 @@ PublicKey BLS::DHKeyExchange(const PrivateKey& privKey, const PublicKey& pubKey)
 
 void BLS::CheckRelicErrors() {
     if (!core_get()) {
-        throw std::string("Library not initialized properly. Call BLS::Init()");
```
Comment (PastaPastaPasta) [inline src/bls.cpp:317]: Same as upstream

**Review by PastaPastaPasta — COMMENTED** [review]
All of these below are the same as upstream. We should submit PR doing this change

**Review by UdjinM6 — APPROVED** [review]
I think we can fix it here and submit a similar PR to upstream instead of waiting for upstream to fix this and backport from v2.

utACK

**Comment by PastaPastaPasta** [comment]
I'd like a PR to be opened upstream (doesn't need to be merged upstream) before we merge this PR

**Comment by xdustinface** [comment]
See https://github.com/Chia-Network/bls-signatures/pull/147

**Review by PastaPastaPasta — APPROVED** [review]
utACK

**Pull request merged**